File: service.go
package spine

import (
	"context"
	"fmt"
	"io"

	"github.com/grandcat/zeroconf"
	"github.com/poisnoir/mad-go"
	"github.com/xtaci/kcp-go/v5"
)

type Service[K any, V any] struct {
	namespace *Namespace
	name      string
	server    *zeroconf.Server

	keySerializer   *mad.Mad[K]
	valueSerializer *mad.Mad[V]
	errorSerializer *mad.Mad[string]

	context  context.Context
	listener *kcp.Listener
	cancel   context.CancelFunc

	handler  func(K) (V, error)
	requests chan serviceRequest[K, V]
}

func NewService[K any, V any](namespace *Namespace, name string, handler func(K) (V, error)) (*Service[K, V], error) {

	// fix me pls
	logger := namespace.logger

	keyEnc, valueEnc, listener, server, err := generateService[K, V](namespace, name)
	if err != nil {
		return nil, fmt.Errorf("failed to create service: %v", err)
	}

	errEnc, _ := mad.NewMad[string]()
	ctx, cancel := context.WithCancel(namespace.ctx)

	s := &Service[K, V]{
		namespace: namespace,
		name:      name,
		server:    server,

		keySerializer:   keyEnc,
		valueSerializer: valueEnc,
		errorSerializer: errEnc,

		context:  ctx,
		cancel:   cancel,
		listener: listener,

		requests: make(chan serviceRequest[K, V], 100),
		handler:  handler,
	}

	go s.runHandler()
	go runListener(listener, logger, s.clientHandler) // stops when listener closes
	return s, nil
}

func (s *Service[K, V]) clientHandler(conn io.ReadWriteCloser) {

	logger := s.namespace.logger.With(
		s.namespace.Name(),
		"service",
		s.name,
		"client handler",
	)

	bufPtr := s.namespace.bufferPool.Get().(*[]byte)
	defer s.namespace.bufferPool.Put(bufPtr)

	handleCallerRequest(conn, s.keySerializer, s.valueSerializer, s.errorSerializer, *bufPtr, s.processRequest, logger)

}

func (s *Service[K, V]) processRequest(key K) serviceOutput[V] {
	// send to handler
	hr := serviceRequest[K, V]{
		input:  key,
		output: make(chan serviceOutput[V], 1),
	}

	// todo: need some timeout shit
	s.requests <- hr
	return <-hr.output
}

func (s *Service[K, V]) runHandler() {

	logger := s.namespace.logger.With(
		s.namespace.Name(),
		"service",
		s.name,
		"request handler",
	)

	for {
		select {
		case request := <-s.requests:
			response, err := s.handler(request.input)
			if err != nil {
				logger.Error("unable to handle request", "error", err)
				// Todo: Change To return error instead of default
			}
			request.output <- serviceOutput[V]{data: response, err: err}
			s.namespace.logger.Info("handled request", "request", request, "response", response)
		case <-s.context.Done():
			return
		}
	}
}

func (s *Service[K, V]) Close() {
	s.cancel()
}

func (s *Service[K, V]) Name() string {
	return s.name
}

File: py/service.py
import asyncio
import logging
from .service_common import generate_service, handle_caller_request
from .mad import new_mad
from . import globals

class Service:
    def __init__(self, namespace, name, handler):
        self.name = name
        self.namespace = namespace
        self.handler = handler
        self.logger = namespace.logger.getChild(f"service.{name}")
        self.requests = asyncio.Queue(maxsize=100)
        self.ctx = asyncio.Event()
        self.tasks = []
        
        # Encoders
        self.key_encoder = None
        self.value_encoder = None
        self.error_encoder = new_mad("string")
        self.info = None

    async def start(self):
        self.key_encoder, self.value_encoder, self.info = await generate_service(self.namespace, self.name)
        
        # Start handler loop
        self.tasks.append(asyncio.create_task(self.run_handler()))
        
        # Start listener
        # In actual KCP library, we'd start a KCP server here
        # self.tasks.append(asyncio.create_task(self.run_listener()))
        
    async def run_handler(self):
        while not self.ctx.is_set():
            try:
                request_data, output_queue = await self.requests.get()
                try:
                    # Run handler
                    response = await self.handler(request_data)
                    await output_queue.put((response, None))
                except Exception as e:
                    self.logger.error(f"Handler error: {e}")
                    await output_queue.put((None, e))
                finally:
                    self.requests.task_done()
            except asyncio.CancelledError:
                break
            except Exception as e:
                self.logger.error(f"Handler loop error: {e}")

    async def process_request(self, key):
        output_queue = asyncio.Queue(maxsize=1)
        await self.requests.put((key, output_queue))
        return await output_queue.get()

    async def client_handler(self, reader, writer):
        await handle_caller_request(reader, writer, self.key_encoder, self.value_encoder, self.process_request, self.logger)

    async def close(self):
        self.ctx.set()
        for task in self.tasks:
            task.cancel()
        if self.info:
            self.namespace.reg.zc.unregister_service(self.info)

def new_service(namespace, name, handler):
    service = Service(namespace, name, handler)
    # Note: need to await service.start() in asyncio context
    return service
